# Checkpoint 1 - Gráficos, Estatísticas e Análise Exploratória

### Instruções:

- Neste Jupyter Notebook há 10 questões de Análise Exploratória. Cada questão possui uma pontuação indicada em vermelho, somando 10 pontos ao final;
- O CP é em grupo, o mesmo grupo do Challenge;
- Preencham o nome de cada membro do grupo corretamente;
- Coloque seu código solução logo abaixo de cada questão que ele responde;
- Use células markdown para escrever respostas que forem pedidas ou que você achar pertinentes;
- Você pode e **deve** usar bibliotecas prontas. Ao importar uma biblioteca, comente na frente para o que ela serve;
- Antes de submeter o jupyter no Teams, renomeie-o para **CP1-turma-NomeSobrenome.ipynb** onde **turma** deve ser trocado pela turma do grupo (H,A,J,F,G) e o NomeSobrenome trocado pelo nome e sobrenome do aluno que submeter o arquivo.

### Atenção:
Vocês podem usar trechos de códigos prontos, entretanto é pouco provável que dois grupos tenham códigos **totalmente iguais**, nas mesmas sequências com as mesmas respostas. Caso for constatada semelhança exagerada entre dois arquivos que indique **plágio**, a nota  <font color='red'> será zerada!</font>

## Grupo [nome do grupo] Turma: H
- **Nome:** Bruno Lora **RM:89321**
- **Nome:** Matheus Meneguim **RM:88123**
- **Nome:** Pedro Leite **RM:86940**

------------------------------

## Fifa 2019

No arquivo data_fifa2019.csv há dados de jogadores de futebol inseridos no jogo eletrônico Fifa.

Vocês devem realizar aquilo que é pedido em cada Exercício.

Mais informações sobre esse conjunto de dados pode ser obtida aqui [Fifa 2019](https://www.kaggle.com/karangadiya/fifa19).

### 1) Importe as bibliotecas necessárias e carregue os dados do arquivo 'data_fifa2019.csv' em um pandas dataframe e respoda <font color='red'>(0.5 ponto)</font>:

(a) Quantos jogadores existem nessa base da dados?

(b) Quantos atributos de cada jogador são avaliados?

(c) Quantos desses atributos são numéricos e quantos são categóricos?

In [63]:
import pandas as pd

In [64]:
import statistics

In [65]:
import matplotlib.pyplot as plt  # Para trabalhar com gráficos
import seaborn as sns

ModuleNotFoundError: No module named 'matplotlib'

In [66]:
dados = pd.read_csv('data_fifa2019.csv',sep=',')

In [67]:
#o numero de linhas equivale ao numero de jogadores
linhas = len(dados)

print('(a) Quantos jogadores existem nessa base da dados?')
print('(b) Quantos atributos de cada jogador são avaliados?\n')
print('Numero de jogadores na tabela: ',linhas,'linhas')

#o numero de colunas equivale ao numero de atributos
colunas = (len(dados.columns))
print('Numero de atributos na tabela: ',colunas,'colunas')


#converte os dados para dtypes, melhor selecao dos valores
dados_convertidos = dados.convert_dtypes()

#print dos dados convertidos
#dt=(dados_convertidos.dtypes)
#print(dt)

print('\n(c) Quantos desses atributos são numéricos e quantos são categóricos? \n')
#pega os valores Int64 e Float64 dentro dos dados convertidos
numeros_tabela = (dados_convertidos.select_dtypes(include=['Int64']))

#pega valores que nao sao considerados Int64 dos dados convertidos
nao_numeros_tabela = (dados_convertidos.select_dtypes(exclude=['Int64']))

#printa o numero de colunas em numeros_tabela
atributos_numericos = len(numeros_tabela.columns)
print('Numero de atributos numericos: ',atributos_numericos,'colunas')

#printa o numero de colunas em nao numeros tabela
atributos_nao_numericos = len(nao_numeros_tabela.columns)
print('Numero de atributos categoricos: ',atributos_nao_numericos,'colunas')

(a) Quantos jogadores existem nessa base da dados?
(b) Quantos atributos de cada jogador são avaliados?

Numero de jogadores na tabela:  18207 linhas
Numero de atributos na tabela:  89 colunas

(c) Quantos desses atributos são numéricos e quantos são categóricos? 

Numero de atributos numericos:  44 colunas
Numero de atributos categoricos:  45 colunas


####  a)O número de jogadores nessa base de dados é 18207 jogadores.

####  b)O número de atributos para cada jogador é 89 atributos

####  c)45 desses atributos são categóricos e 44 são numéricos

### 2) Faça uma função para arrumar os valores da coluna Value dos jogadores. <font color='red'>(0.75 ponto)</font>

Você deve fazer uma função que recebe uma string e converte a string para um float, que representa o valor do jogador em euros. Lembre que **K** significa 1000 e **M** significa 1000000. O símbolo de euro é **€**. Não esqueça de forçar o tipo float.

Após criar a função, aplique-a a coluna de dados 'value' do seu dataframe e obtenha em uma coluna o valor de cada jogador.

In [12]:
def arrumar_value(x):
    dados_str = x
    value_float = float(dados_str)
    return value_float

In [13]:
nome = dados['Value'].apply(lambda x: str(x.replace("€","")))

In [14]:
dados['Value'] = dados['Value'].apply(arrumar_value)

ValueError: could not convert string to float: '€110.5M'

In [33]:
print(dados)

       Unnamed: 0      ID                Name  Age  \
0               0  158023            L. Messi   31   
1               1   20801   Cristiano Ronaldo   33   
2               2  190871           Neymar Jr   26   
3               3  193080              De Gea   27   
4               4  192985        K. De Bruyne   27   
...           ...     ...                 ...  ...   
18202       18202  238813        J. Lundstram   19   
18203       18203  243165  N. Christoffersson   19   
18204       18204  241638           B. Worman   16   
18205       18205  246268      D. Walker-Rice   17   
18206       18206  246269           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4     

### 3) Usando a função overall_classes, crie uma coluna denominada 'Performance' com as classes de jogadores baseado no valor da coluna 'Overall'. Gere uma distribuição (histograma) com a Performance de todos os jogadores e outra com a coluna Overall. Compare as duas. <font color='red'>(0.75 ponto)</font>

In [15]:
# Criando três classes baseado no Overall
# A função recebe um valor e retorna uma classe categórica da performance do jogador (Low, Normal, High)
def overall_classes(x):
    if x < 60:
        return 'Baixa'
    if x < 75:
        return 'Normal'
    return 'Alta'

In [16]:
dados['Performance'] = dados['Overall'].apply(overall_classes)

In [17]:
fig, axes = plt.subplots(1, 2, figsize=(10, 4))
fig.suptitle('Dados de Performance e Overall')


sns.histplot(ax=axes[0],
            data=dados['Overall'],
            kde=True,
            bins=10,
            linewidth=0.7,
            )

sns.histplot(ax=axes[1],
            data=dados['Performance'],
            kde=True,
            bins=10,
            linewidth=0.7,
            )

axes[0].set(xlabel='Overall', ylabel='Jogadores')
axes[1].set(xlabel='Performance', ylabel='Jogadores')

NameError: name 'plt' is not defined

####  Resposta: A comparação dos dois histogramas demonstra que eles indicam o mesmo resultado, sendo que uma parte significativa dos atletas tem uma performance normal e tem um Overall entre 60 e 75. Já as outras são menos significativas, sendo que poucos atletas estão acima ou abaixo da média.

### 4) Em relação a idade (Age) dos jogadores, calcule a média aritmética e a incerteza da média: <font color='red'>(1 ponto)</font>

(a) Todos os jogadores

(b) Jogadores de Alta Performance

(c) Jogadores de Performance Normal

(d) Jogadores de Baixa Performance

In [18]:
# Variavel recebe todos as age dos jogadores
age = (dados['Age'])
#performance dos jogadores
performance = (dados['Performance'])

print('(a) Todos os jogadores: \n')
#media aritmetica da Age de todos os jogadores
media_todos = round(age.sum()/len(age), 2)
print('Media das ages, todos os jogadores:',media_todos)

# incerteza da media de todos os jogadores
N = len(age)
media = media_todos

soma=0
for xi in age:
    soma = soma+(xi-media)**2
incerteza_todos = (soma/(N*(N-1)))**(1/2)
print('Incerteza da media, todos os jogadores',incerteza_todos, '\n')



print('(b) Jogadores de Alta Performance:\n')
#pega os jogadores de Alta Performance
performance_alta = dados[performance.str.contains('Alta')]

#pega Age dos jogadores com alta performance
age_perf_alta = (performance_alta['Age'])

#media aritmetica da Age dos jogadores de Alta Performance
media_AltaP = age_perf_alta.sum() / len(age_perf_alta)
print('Media das ages, jogadores de Alta Performance:', media_AltaP)

#incerteza da media dos jogadores de Alta performance
N = len(age_perf_alta)
media = media_AltaP

soma=0
for xi in age_perf_alta:
    soma = soma+(xi-media)**2
incerteza_AltaP = (soma/(N*(N-1)))**(1/2)
print('Incerteza da media, jogadores Alta Performance:',incerteza_AltaP, '\n')



print('(c) Jogadores de Performance Normal:\n')
#pega os jogadores de Performance Normal
performance_normal = dados[performance.str.contains('Normal')]

#pega Age dos jogadores com Performance Normal
age_perf_normal = (performance_normal['Age'])

#media aritmetica da Age dos jogadores de Performance Normal
media_PNormal = age_perf_normal.sum() / len(age_perf_normal)
print('Media das ages, jogadores de Performance Normal:', media_PNormal)

#incerteza da media dos jogadores de Performance Normal
N = len(age_perf_normal)
media = media_PNormal

soma=0
for xi in age_perf_normal:
    soma = soma+(xi-media)**2
incerteza_PNormal = (soma/(N*(N-1)))**(1/2)
print('Incerteza da media, jogadores Performance Normal:',incerteza_PNormal, '\n')


print('(d) Jogadores de Baixa Performance:\n')
#pega os jogadores de Performance Baixa
performance_baixa = dados[performance.str.contains('Baixa')]

#pega Age dos jogadores com Performance Baixa
age_perf_baixa = (performance_baixa['Age'])

#media aritmetica da Age dos jogadores de Performance Baixa
media_PBaixa = age_perf_baixa.sum() / len(age_perf_baixa)
print('Media das ages, jogadores de Performance Baixa:', media_PBaixa)

#incerteza da media dos jogadores de Performance Baixa
N = len(age_perf_baixa)
media = media_PBaixa

soma=0
for xi in age_perf_baixa:
    soma = soma+(xi-media)**2
incerteza_PBaixa = (soma/(N*(N-1)))**(1/2)
print('Incerteza da media, jogadores Performance Normal:',incerteza_PBaixa, '\n')


(a) Todos os jogadores: 

Media das ages, todos os jogadores: 25.12
Incerteza da media, todos os jogadores 0.03460926732466789 

(b) Jogadores de Alta Performance:

Media das ages, jogadores de Alta Performance: 27.244197292069632
Incerteza da media, jogadores Alta Performance: 0.08488710046240193 

(c) Jogadores de Performance Normal:

Media das ages, jogadores de Performance Normal: 25.73146958821307
Incerteza da media, jogadores Performance Normal: 0.03861406249112684 

(d) Jogadores de Baixa Performance:

Media das ages, jogadores de Performance Baixa: 20.834366391184574
Incerteza da media, jogadores Performance Normal: 0.06666424402621199 



### 5) Faça um gráfico de barras com as médias calculadas no exercício 4. Inclua as incertezas como barras de erro associadas a cada barra de valor. Você terá que olhar na documentação do matplotlib! <font color='red'>(1 ponto)</font>

In [19]:
#valores das medias - eixo X
medias_gerais = ['Geral', 'Alta', 'Normal', 'Baixa']
#nomes das colunas - eixo Y
valores_gerais = [media_todos, media_AltaP, media_PNormal, media_PBaixa]

#Inserindo os valores anteriores para o plt.bar
plt.bar=(medias_gerais, valores_gerais)

#criando uma figure, axes, alterando tamanho
fig, ax = plt.subplots(figsize=(5,6))

#criando o gráfico de barras
sns.barplot(x=medias_gerais, y=valores_gerais, ax=ax, data=dados, palette='RdPu')

#adicionando título
ax.set_title("Media Rendimento X Performance Jogadores", fontdict={'fontsize':15})

#mudando nome e tamanho do label x
ax.set_xlabel('Performance dos jogadores', fontdict={'fontsize':14})

#mudando nome e tamanho do label y
ax.set_ylabel('Media de Rendimentos', fontdict={'fontsize':14})

#mudando tamanho dos labels dos ticks
ax.tick_params(labelsize=14)

#otimizar espaço da figure
fig.tight_layout();


#valor das incertezas
valor_incertezas = [incerteza_todos, incerteza_AltaP, incerteza_PNormal, incerteza_PBaixa]
#nome das incertezas
nome_incertezas = ['Geral', 'Alta', 'PNormal', 'PBaixa']

#inserindo os valores anteriores para o plt.errorbar
plt.errorbar(medias_gerais, valores_gerais, yerr = valor_incertezas)

NameError: name 'plt' is not defined

### 6) Repita os exerícios 4 e 5 fazendo para a coluna de valor (Value) dos jogadores. <font color='red'>(0.75 ponto)</font>

In [20]:
# Coloque seu código aqui

### 7) Faça um gráfico boxplot da idade e do valor dos jogadores onde cada boxplot é uma classe de performance. O que você pode interepretar desse gráfico? <font color='red'>(1 ponto)</font>


In [21]:
fig, axes = plt.subplots(1, 2, figsize=(15, 80))
fig.suptitle('Dados dos Atletas')

sns.boxplot(x="Age", y="Value",hue="Performance",
            palette=["g","r","b"],
            linewidth=0.9,
            ax=axes[0],
            data=dados
           )


NameError: name 'plt' is not defined

####  Resposta: É possível interpretar pelo gráfico que o valor dos jogadores está diretamente relacionado com a performance. A idade pode afetar a performance mas isso é mais observado nos jogadores mais novos que, geralmente tem uma performance "baixa".


### 8) Faça um gráfico de dispersão entre idade e valor. Considere gerar um gráfico para todos os jogadores, e 3 gráficos para cada classe de performance. Existe algum padrão? <font color='red'>(1 ponto)</font>

Dica: use o eixo de valor em escala logarítmica.

In [22]:
sns.jointplot(data=dados,x="Age", y="Value",
              height=30, 
              ratio=2,
              marginal_ticks=True,
              hue = 'Performance'
             )

NameError: name 'sns' is not defined

####  Resposta: É possível interpretar pelo gráfico que o valor dos jogadores está relacionado com a performance do individuo. Os jogadores com performance baixa tem valores mais baixos e um número significativo deles tem menos de 26 anos enquanto que os jogadores com performance média e alta tem uma maior distribuição de idades.

### 9) Em relação a nacionalidade e ao clube dos jogadores, responda: <font color='red'>(2 pontos)</font>

(a) Qual é o país com o maior número de jogadores na classe de Alta Performance? E de performance Normal e Baixa?

(b) Qual é o país com maior média de idade dos jogadores? E qual é o país com menor média?

(c) Qual é o clube com o maior número de jodarores na classe de Alta Performance?

(d) Qual é o clube mais valioso em termos de valor de jogadores? E o menos valioso? Qual é a média e desvio padrão amostral de valor dos clubes?

(e) Plote um histograma do valor dos clubes (soma do valor dos jogadores do clube); 

**Dica:** Use a função groupby do Pandas encadeada com funções de média, soma e contagem (mean, sum, count).

In [23]:
len(set(dados['Nationality'])) # Número de paises

164

In [57]:
# #pega Nacionalidade dos jogadores com alta performance
nat_alta = (performance_alta['Nationality'])

print(nat_alta.value_counts())


# #pega Nacionalidade dos jogadores com normal performance
nat_media = (performance_normal['Nationality'])

print(nat_media.value_counts())

# #pega Nacionalidade dos jogadores com baixa performance
nat_baixa = (performance_baixa['Nationality'])

print(nat_baixa.value_counts())

Spain                   260
Brazil                  203
France                  165
Argentina               147
Germany                 128
                       ... 
China PR                  1
United Arab Emirates      1
Kenya                     1
Equatorial Guinea         1
Northern Ireland          1
Name: Nationality, Length: 97, dtype: int64
England           1096
Germany            900
Spain              773
Argentina          722
France             638
                  ... 
St Kitts Nevis       1
Guam                 1
Kuwait               1
Mauritius            1
Uzbekistan           1
Name: Nationality, Length: 161, dtype: int64
England                463
China PR               178
Germany                170
Republic of Ireland    157
Japan                  138
                      ... 
Antigua & Barbuda        1
Philippines              1
Guyana                   1
Tunisia                  1
Bolivia                  1
Name: Nationality, Length: 106, dtype: int64


In [35]:
print(dados['Nationality'])

0        Argentina
1         Portugal
2           Brazil
3            Spain
4          Belgium
           ...    
18202      England
18203       Sweden
18204      England
18205      England
18206      England
Name: Nationality, Length: 18207, dtype: object


In [76]:
print(age.groupby([nacionalidade]).sum())

Nationality
Afghanistan      90
Albania         961
Algeria        1623
Andorra          28
Angola          388
               ... 
Uzbekistan       59
Venezuela      1629
Wales          3191
Zambia          200
Zimbabwe        349
Name: Age, Length: 164, dtype: int64


AttributeError: 'NoneType' object has no attribute 'max'

In [58]:
media_nt = (media_todos['Nationality'])

print(media_nt.value_counts())

IndexError: invalid index to scalar variable.

In [59]:
#pega o Clube  com alta performance
clube_alta = (performance_alta['Club'])
print(clube_alta.value_counts())

Chelsea                  24
Manchester United        23
Tottenham Hotspur        23
Liverpool                23
Napoli                   23
                         ..
Kaizer Chiefs             1
Junior FC                 1
Palermo                   1
Puebla FC                 1
1. FC Heidenheim 1846     1
Name: Club, Length: 290, dtype: int64


### 10) Pergunta aberta <font color='red'>(1.25 pontos)</font>:

Após analisar todos os dados, você conseguiria traçar uma estratégia de investimento para lucrar com algum time? 

Quais são as características dos times de alta performance?

Você acha possível analisar usando apenas idade e valor, ou precisaria usar outros dados da base da dados disponível?

####  Resposta:Analisando times que possuem muitos atletas de alta performance é interessante investir pois esse time é mais provável de ter uma boa performance em competições.

####  Resposta: Clubes de alta performance como por exemplo: Chelsea, Man United, Liverpool... são clubes que sempre estão nas melhores competições e ganhando as premiações, que juntamente com os patriocínios podem acabar investindo em promessas de jogadores jovens que futuramente podem se tornar estrelas, ou até mesmo comprando jogadores de outros clubes, fazendo assim o clube crescer e ficar mais famoso.

####  Resposta:Sim, é possível analisar apenas com idade e valor, observa-se que os jogadores mais jovens geralmente não tem um valor tão alto mas que jogadores entre 24 anos e 35 anos tem um valor mais alto. Com mais dados conseguiriamos entender quais fatores afetam o valor e porque os atletas mais jovens costumam ter um valor mais baixo.